proposals/6/README.md

```diff
@@ -0,0 +1,123 @@
+  * Correspondingly avoid joining API resources - at least when listing them.
+* Security: Don't mix security concerns. Think of how different security-related responsibilities can be separated before implementing new features. Model CRDs so that K8s' RBAC can be leveraged.
+
+## 4. External interface(s) & user load
+
+While Jenkins X cluster-internal process integrations should be done using the Kubernetes API directly
+this shouldn't be done unnecessarily with clients outside the cluster because
+* Companies may not want to expose their K8s API.
+* Exposing k8s API - even if well secured - results in a larger attack surface.
+* Clients that rely on k8s API resources must evolve with every (partial) k8s API/CRD change.
+* User load should ideally not affect the K8s API server since it is usually not scaled as much as regular cluster workloads and the API server resources must be kept free for cluster-internal operations.
+
+Therefore, if there is no particular need to access the k8s API, an abstract domain-specific middleware API (e.g. JXUI backend) should be used by external clients as a gateway into the cluster (user requests should be served from memory as much as possible using CQRS).
+The middleware API also serves as a system boundary between cluster-internal logic/APIs and external/client-side tools (e.g. GUI, CLI) allowing us to evolve both independent from each other.
+This way whenever an API version inside the cluster changes (deprecation?!) only the middleware has to be adjusted - not all clients.
+Exposing only such a specialized middleware outside the cluster also reduces the attack surface.
+
+## 5. Action plan
+
+1. Discuss, agree on or amend these principles.
+1. Apply them while decomposing jx ([proposal #5](../5/README.md)).
+1. Apply them when designing new features.
+
+## 6. Potential implementations in Jenkins X
+
+Not everything needs to be implemented as CRD and Kubernetes controller.
+However it is beneficial for 3rd party integrations and our own experimentation if a CRD or k8s event is used to trigger each automated business process since this allows to decouple components and 3rd parties to plug in / trigger their own components as a replacement or addition of Jenkins X' components and is an idiom of the k8s platform.
+
+TODO: To be refined:
+* SourceRepository controller that accepts pipeline bot invitations (more of a security concern) - maybe not a good idea after all since e.g. github request for bot invitations could hang and therefore block the reconcile loop -> would need to spawn a pod to accept the invite
+* Promotion controller that triggers a rollout and reflects its state in the corresponding CR's status so that clients (like jxui) can easily query it) - question remains how to make this work consistently with GitOps
+* ...
```

Comment:
I believe that RoleController is currently being extracted out of JX and there is also work happening on an Environment/Preview Controller.